pkg/reconciler/namespace/namespace.go

```diff
@@ -0,0 +1,210 @@
+/*
+Copyright 2021 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package namespace
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/clusters"
+	"k8s.io/klog/v2"
+
+	"github.com/kcp-dev/kcp/pkg/apis/cluster/v1alpha1"
+)
+
+const clusterLabel = "kcp.dev/cluster"
+
+// reconcileResource is responsible for setting the cluster for a resource of
+// any type, to match the cluster where its namespace is assigned.
+func (c *Controller) reconcileResource(ctx context.Context, lclusterName string, unstr *unstructured.Unstructured, gvr *schema.GroupVersionResource) error {
+	klog.Infof("Reconciling %s %s/%s", gvr.String(), unstr.GetNamespace(), unstr.GetName())
+
+	// If the resource is not namespaced (incl if the resource is itself a
+	// namespace), ignore it.
+	if unstr.GetNamespace() == "" {
+		klog.Infof("%s %s had no namespace; ignoring", gvr.String(), unstr.GetName())
+		return nil
+	}
+
+	// Align the resource's assigned cluster with the namespace's assigned
+	// cluster.
+	// First, get the namespace object (from the cached lister).
+	ns, err := c.namespaceLister.Get(clusters.ToClusterAwareKey(lclusterName, unstr.GetNamespace()))
+	if err != nil {
+		return err
+	}
+
+	lbls := unstr.GetLabels()
+	if lbls == nil {
+		lbls = map[string]string{}
+	}
+
+	old, new := lbls[clusterLabel], ns.Labels[clusterLabel]
+	if old == new {
+		// Already assigned to the right cluster.
+		return nil
+	}
+
+	// Update the resource's assignment.
+	klog.Infof("Patching to update cluster assignment for %s %s/%s: %q -> %q", gvr, ns.Name, unstr.GetName(), old, new)
+	if _, err := c.dynClient.Resource(*gvr).Namespace(ns.Name).Patch(ctx, unstr.GetName(), types.MergePatchType, clusterLabelPatchBytes(new), metav1.PatchOptions{}); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *Controller) reconcileGVR(ctx context.Context, gvr schema.GroupVersionResource) error {
+	// Update all resources in the namespace with the cluster assignment.
+	listers, _ := c.ddsif.Listers()
+	lister, found := listers[gvr]
+	if !found {
+		klog.Infof("Informer for %s is not synced; re-enqueueing", gvr)
+		c.enqueueGVR(gvr)
```

Comment:
Maybe return err here instead? - it will bubble up and get requeued w/rate limiting

Reply:
Good call, done.